components/raftstore-v2/src/operation/query/replica.rs

```diff
@@ -0,0 +1,161 @@
+// Copyright 2022 TiKV Project Authors. Licensed under Apache-2.0.
+
+use engine_traits::{KvEngine, RaftEngine};
+use kvproto::raft_cmdpb::{self, CmdType, RaftCmdRequest, RaftCmdResponse};
+use pd_client::INVALID_ID;
+use raftstore::{
+    store::{
+        cmd_resp,
+        fsm::apply::notify_stale_req,
+        metrics::RAFT_READ_INDEX_PENDING_COUNT,
+        msg::{ErrorCallback, ReadCallback},
+        propose_read_index,
+        util::check_region_epoch,
+        ReadIndexRequest, Transport,
+    },
+    Error,
+};
+use slog::{debug, error, info, o, Logger};
+use tikv_util::{box_err, time::monotonic_raw_now};
+use time::Timespec;
+use tracker::GLOBAL_TRACKERS;
+
+use crate::{
+    batch::StoreContext,
+    raft::Peer,
+    router::{message::RaftRequest, QueryResChannel},
+    Result,
+};
+impl<EK: KvEngine, ER: RaftEngine> Peer<EK, ER> {
+    /// read index on follower
+    ///
+    /// return true if it's proposed.
+    pub(crate) fn read_index_follower<T: Transport>(
+        &mut self,
+        poll_ctx: &mut StoreContext<EK, ER, T>,
+        mut req: RaftCmdRequest,
+        ch: QueryResChannel,
+    ) {
+        let now = monotonic_raw_now();
+        if self.leader_id() == INVALID_ID {
+            poll_ctx
+                .raft_metrics
+                .invalid_proposal
+                .read_index_no_leader
+                .inc();
+            let mut err_resp = RaftCmdResponse::default();
+            let term = self.term();
+            cmd_resp::bind_term(&mut err_resp, term);
+            cmd_resp::bind_error(&mut err_resp, Error::NotLeader(self.region_id(), None));
+            ch.report_error(err_resp);
+            return;
+        }
+
+        if self.propose_read_index(poll_ctx, req, self.is_leader(), ch, now) {
```

Comment:
I suggest to inline the `propose_read_index` implementation to make it clear. Otherwise it's weird to let replica use the same method from lease module and lease module needs to check whether it's leader.